File: shared/common/src/main/java/cafe/jeffrey/shared/common/model/repository/RepositoryStatistics.java
/*
 * Jeffrey
 * Copyright (C) 2026 Petr Bouda
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package cafe.jeffrey.shared.common.model.repository;

/**
 * Aggregate figures about a project's recording repository.
 *
 * <p>Totals only, on purpose. This used to carry a count and a size for each of six buckets — JFR,
 * heap dump, log, app log, error log, other — filled by a switch over {@link ManagedFile}. That
 * switch was the last place anything on the hub side decided what a file type <em>means</em>, and
 * it decided wrong: pprof and OTLP recordings landed in "other", beside the files nothing could
 * classify at all.
 *
 * <p>Nothing here names a kind of file, so adding a {@code ManagedFile} constant changes neither
 * this record nor the RPC that carries it. A caller that wants to know what a particular file is
 * reads {@code fileType} off the session listing, where the hub reports it without acting on it.
 */
public record RepositoryStatistics(
        int totalSessions,
        RecordingStatus latestSessionStatus,
        long lastActivityTimeMillis,
        long totalSizeBytes,
        int totalFiles,
        long biggestSessionSizeBytes) {

    public static final RepositoryStatistics EMPTY =
            new RepositoryStatistics(0, RecordingStatus.UNKNOWN, 0L, 0L, 0, 0L);
}

File: shared/ui/hubs/ui/services/api/model/RepositoryStatistics.ts
/*
 * Jeffrey
 * Copyright (C) 2024 Petr Bouda
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import RecordingStatus from '@hubs/services/api/model/RecordingStatus.ts';

/**
 * A project's repository in totals. There is deliberately no breakdown by file type: the hub used
 * to send six fixed buckets, which meant deciding what every type means, and a pprof or OTLP
 * recording arrived as "other". Adding a file type now changes nothing here.
 */
export default interface RepositoryStatistics {
  // Session Overview
  totalSessions: number;
  sessionStatus: RecordingStatus; // Status of the latest session
  lastActivityTime: number; // Timestamp in milliseconds - for use with FormattingService.formatRelativeTime()

  // Storage Overview
  totalSize: number; // Total repository size in bytes
  totalFiles: number; // Total number of files across all sessions
  biggestSessionSize: number; // Size of the largest session in bytes
}
